src/sqlfluff/core/dialects/dialect_ansi.py

```diff
@@ -1100,6 +1100,10 @@ class FromClauseSegment(BaseSegment):
         AnyNumberOf(
             Ref("JoinClauseSegment"), Ref("JoinLikeClauseGrammar"), optional=True
         ),
+        # In case of table functions e.g. lateral flatten
+        AnyNumberOf(
+            Sequence(Ref("CommaSegment"), Ref("TableExpressionSegment")), optional=True
```

Comment:
I'm not sure this is the best way to do this!

Reply:
This _works_, but isn't where I would put it. I would put it within the `Delimited(OneOf(...))` in the `parse_grammar` of `FromClauseSegment`. That's the same place which deals with the old skool join syntax of `SELECT * FROM a, b WHERE ...`.

Reply:
Ah so the `TableExpressionSegment` already exists there:

https://github.com/sqlfluff/sqlfluff/pull/791/files#diff-e07ae5ab00e7f519401a3251b192139d2da4ebf08fd1d7403b795472bff4f29aR1089

But the issue is the terminator being a `JoinClauseSegment`:

https://github.com/sqlfluff/sqlfluff/pull/791/files#diff-e07ae5ab00e7f519401a3251b192139d2da4ebf08fd1d7403b795472bff4f29aR1091

When we can have lateral flattens after a join as demonstrated in the added test case https://github.com/sqlfluff/sqlfluff/pull/791/files#diff-9b90add4f247b50c3d7d4bf0b683ed7b96af8a2569a87fafc61a287ca6a979da

I'll have a think

Reply:
I think I've fixed this by refactoring the `FromClauseSegment` and `TableExpressionSegment`. From clauses now live under `TableExpressionSegment`s, which in my head feels right and also according these docs: https://www.cockroachlabs.com/docs/v20.2/sql-grammar.htm#table_ref

What do you reckon?